File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/Blob.kt
package com.jetbrains.signatureverifier.serialization

import com.jetbrains.signatureverifier.macho.CSMAGIC
import kotlinx.serialization.Serializable

@Serializable
data class Blob(
  val type: UInt = 0u,
  val offset: UInt = 0u,
  val magic: CSMAGIC = CSMAGIC.UNKNOWN,
  val magicValue: UInt = 0u,
  var length: Int = 0,
  @Serializable(ByteArraySerializer::class)
  var content: ByteArray = byteArrayOf()
) {
  constructor(type: UInt, offset: UInt, magic: CSMAGIC, magicValue: UInt, content: ByteArray) : this(
    type, offset, magic, magicValue, content.size, content
  )

  fun toByteArray() = magicValue.toByteArray(true) + length.toByteArray(true) + content
}

File: net/JetBrains.FormatRipper/src/Dmg/Blob.cs
using JetBrains.FormatRipper.Impl;

namespace JetBrains.FormatRipper.Dmg;

public class Blob
{
  public readonly uint Type;
  public readonly uint Offset;
  public readonly CSMAGIC_CONSTS Magic;
  public readonly uint MagicValue;
  public int Length;
  public byte[] Content;

  public Blob(uint type, uint offset, CSMAGIC_CONSTS magic, uint magicValue, byte[] content, int? length = null)
  {
    Type = type;
    Offset = offset;
    Magic = magic;
    MagicValue = magicValue;
    Length = length ?? content.Length;
    Content = content;
  }

  public byte[] ToByteArray() =>
    MemoryUtil.ArrayMerge(
      MemoryUtil.ToByteArray(MagicValue, true),
      MemoryUtil.ToByteArray(Length, true),
      Content
    );
}